### PR TITLE
fix: Don't replace malformed appmap.yml

### LINKED ISSUE #80: Malformed `appmap.yml` is replaced on start
There is a bug that causes the appmap.yml file to be replaced if the agent doesn't understand it. Instead it should show a warning and leave it as-is, using defaults.

### PR DESCRIPTION
**Pull request by zermelo-wisen**
Fixes #80.

### CODE REVIEW

**Review by dividedmind — CHANGES_REQUESTED** [review]

File: src/config.ts [hunk 1]
```diff
@@ -94,11 +96,14 @@ interface ConfigFile {
...
   response_body_max_length?: number;
+  malformed?: boolean;
 }
 
 function readConfigFile(path: string | undefined): ConfigFile | undefined {
   if (!path) return;
-  const config = tryOr(() => YAML.parse(readFileSync(path, "utf8")) as unknown);
+  const fileContent = tryOr(() => readFileSync(path, "utf8"));
+  if (!fileContent) return;
+  const config = tryOr(() => YAML.parse(fileContent) as unknown, { malformed: true });
```
[the remark this answers is not in the document]
Comment (zermelo-wisen) [reply src/config.ts]: Actually, at first, I thought about making it so, but later, I thought it leads to needless changes in surrounding parts and switched to this one. Now I changed it to `ConfigFile | "malformed" | undefined` return.

File: src/config.ts [hunk 2]
```diff
@@ -94,12 +97,16 @@ interface ConfigFile {
   name?: string;
   packages?: Package[];
   response_body_max_length?: number;
+  malformed?: boolean;
```
Comment (dividedmind) [inline src/config.ts:100]: I think you forgot to remove this.

**Review by dividedmind — CHANGES_REQUESTED** [review]
I gave this some more thought, and I think it would be good to expand the scope a bit; in particular, this will still lead to replacing the config file on eg. `EACCES` error, which can be a problem. 

Perhaps `readConfigFile` should just rethrow any exceptions except `ENOENT` (which would return `undefined`). The user can then be warned with the specific error message to make it clearer what's going on.

**Comment by zermelo-wisen** [comment]
> I gave this some more thought, and I think it would be good to expand the scope a bit; in particular, this will still lead to replacing the config file on eg. `EACCES` error, which can be a problem.
> 
> Perhaps `readConfigFile` should just rethrow any exceptions except `ENOENT` (which would return `undefined`). The user can then be warned with the specific error message to make it clearer what's going on.

Ok, If they have a config file besides that they don't want it to be replaced, they don't want to use default parameters, too. So we don't go on with the default config if it throws any exception except `ENOENT`. User will need to resolve the specific error with their config file in order to proceed.

**Comment by dividedmind** [comment]
> > I gave this some more thought, and I think it would be good to expand the scope a bit; in particular, this will still lead to replacing the config file on eg. `EACCES` error, which can be a problem.
> > Perhaps `readConfigFile` should just rethrow any exceptions except `ENOENT` (which would return `undefined`). The user can then be warned with the specific error message to make it clearer what's going on.
> 
> Ok, If they have a config file besides that they don't want it to be replaced, they don't want to use default parameters, too. So we don't go on with the default config if it throws any exception except `ENOENT`. User will need to resolve the specific error with their config file in order to proceed.
> 
> What about exiting when it's malformed too?

Yeah, perhaps you're right. Maybe we should just let the exceptions propagate and catch `ENOENT` only in `Config` constructor. I'm starting to think that will be the least surprising. What do you think?

**Comment by zermelo-wisen** [comment]
> > > I gave this some more thought, and I think it would be good to expand the scope a bit; in particular, this will still lead to replacing the config file on eg. `EACCES` error, which can be a problem.
> > > Perhaps `readConfigFile` should just rethrow any exceptions except `ENOENT` (which would return `undefined`). The user can then be warned with the specific error message to make it clearer what's going on.
> > 
> > 
> > Ok, If they have a config file besides that they don't want it to be replaced, they don't want to use default parameters, too. So we don't go on with the default config if it throws any exception except `ENOENT`. User will need to resolve the specific error with their config file in order to proceed.
> > What about exiting when it's malformed too?
> 
> Yeah, perhaps you're right. Maybe we should just let the exceptions propagate and catch `ENOENT` only in `Config` constructor. I'm starting to think that will be the least surprising. What do you think?

I agree. One point that makes me think is that the user should be given the opportunity to run with defaults if they struggle to fix their config file. For this, maybe we should inform them to remove (backup) or rename their config file to run with defaults. Or we can introduce a flag like `--ignore-config-file` or `--use-defaults`, and let the user know they can use these.

**Comment by dividedmind** [comment]
> > > > I gave this some more thought, and I think it would be good to expand the scope a bit; in particular, this will still lead to replacing the config file on eg. `EACCES` error, which can be a problem.
> > > > Perhaps `readConfigFile` should just rethrow any exceptions except `ENOENT` (which would return `undefined`). The user can then be warned with the specific error message to make it clearer what's going on.
> > > 
> > > 
> > > Ok, If they have a config file besides that they don't want it to be replaced, they don't want to use default parameters, too. So we don't go on with the default config if it throws any exception except `ENOENT`. User will need to resolve the specific error with their config file in order to proceed.
> > > What about exiting when it's malformed too?
> > 
> > 
> > Yeah, perhaps you're right. Maybe we should just let the exceptions propagate and catch `ENOENT` only in `Config` constructor. I'm starting to think that will be the least surprising. What do you think?
> 
> I agree. One point that makes me think is that the user should be given the opportunity to run with defaults if they struggle to fix their config file. For this, maybe we should inform them to remove (backup) or rename their config file to run with defaults. Or we can introduce a flag like `--ignore-config-file` or `--use-defaults`, and let the user know they can use these.

Nah, I don't think the flag is necessary. But I like the idea of explaining the problem and indicating the possible solution, eg. with a message like `Error parsing config file at ${path}: ${error.message}.\nYou can remove the file to use the default configuration.`.

File: src/config.ts [hunk 3]
```diff
@@ -98,7 +98,27 @@ interface ConfigFile {
...
+    if (exn && typeof exn == "object" && "code" in exn && exn.code == "ENOENT") return;
+    throw exn;
+  }
+
+  let config;
+  try {
+    config = YAML.parse(fileContent) as unknown;
+  } catch (exn) {
+    const exnMessage =
+      exn && typeof exn == "object" && "message" in exn && typeof exn.message == "string"
```
Comment (dividedmind) [inline src/config.ts:115]: You can just `assert(isNativeError(exn))` for simplicity :)

File: src/util/tryOr.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
-export default function tryOr<T>(fn: () => T): T | undefined {
+export default function tryOr<T, E = undefined>(fn: () => T, elseReturn?: E): T | E | undefined {
```
Comment (dividedmind) [inline src/util/tryOr.ts:1]: I don't think this change is used anymore.

**Review by dividedmind — APPROVED** [review]

File: src/config.ts (hunk 3 above)
Comment (zermelo-wisen) [reply src/config.ts]: Done

File: src/util/tryOr.ts (hunk 4 above)
Comment (zermelo-wisen) [reply src/util/tryOr.ts]: Reverted

**Pull request merged**